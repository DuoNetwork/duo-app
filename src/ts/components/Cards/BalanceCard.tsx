import {
	Constants as WrapperConstants,
	ICustodianContractAddress
} from '@finbook/duo-contract-wrapper';
import classAIcon from 'images/ClassA_black.png';
import classBIcon from 'images/ClassB_black.png';
import ethIcon from 'images/ethIconB.png';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import {} from 'ts/common/types';
import BalanceInfo from 'ts/components/Cards/BalanceInfo';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardTitle, SRefreshButton } from './_styled';
import ExtendExtraDiv from './ExtendExtraDiv';

interface IProps {
	contractAddress: ICustodianContractAddress;
	locale: string;
	account: string;
	eth: number;
	aToken: number;
	bToken: number;
	mobile?: boolean;
	refresh: () => any;
}

export default class BalanceCard extends React.Component<IProps> {
	public render() {
		const {
			eth,
			aToken,
			bToken,
			account,
			refresh,
			locale,
			mobile,
			contractAddress
		} = this.props;
		return (
			<SCard
				title={
					<SCardTitle>
						{CST.TH_BALANCE[locale].toUpperCase()}{' '}
						<SRefreshButton icon="reload" onClick={refresh} />
					</SCardTitle>
				}
				width={mobile ? '100%' : '590px'}
				margin={mobile ? '20px 0 0 0' : '0 0 0 10px'}
				extra={
					<ExtendExtraDiv
						accountShow={
							account && account !== WrapperConstants.DUMMY_ADDR ? account : 'Unknown'
						}
						account={account}
						locale={locale}
					/>
				}
			>
				<SDivFlexCenter horizontal={!mobile} padding="0 10px">
					<BalanceInfo icon={ethIcon} name={CST.TH_ETH} value={eth} mobile={mobile} />
					<BalanceInfo
						icon={classAIcon}
						name={contractAddress.aToken.code}
						value={aToken}
						mobile={mobile}
					/>
					<BalanceInfo
						icon={classBIcon}
						name={contractAddress.bToken.code}
						value={bToken}
						mobile={mobile}
					/>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
