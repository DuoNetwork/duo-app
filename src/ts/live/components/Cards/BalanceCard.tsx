import * as d3 from 'd3';
import * as React from 'react';
import classAIcon from '../../../../images/ClassA_white.png';
import classBIcon from '../../../../images/ClassB_white.png';
import duoIcon from '../../../../images/Duo_white.png';
import ethIcon from '../../../../images/ethIcon.png';
import * as CST from '../../common/constants';
import { ColorStyles } from '../../common/styles';
import { IBalances, ICustodianStates } from '../../common/types';
import { SDivFlexCenter } from '../_styled';
import { SCard, SCardAssetTag, SCardExtendExtraDiv, SCardTitle } from './_styled';

interface IProps {
	account: string;
	balances: IBalances;
	states: ICustodianStates;
}

const BalanceInfo = (props: {
	icon: string;
	name: string;
	value: number;
	allowance?: number;
	totalSupply?: number;
}) => {
	const { icon, name, value, allowance, totalSupply } = props;
	return (
		<SCardAssetTag value={value}>
			<div className="bg-logo">
				<img src={icon} />
			</div>
			<div className="tag-title">
				<h3>{name}</h3>
			</div>
			<div className="tag-content">
				<div>
					<div>
						<div className={'tag-price'}>{d3.formatPrefix(',.8', 1)(value)}</div>
						{allowance !== undefined ? (
							<div className="tag-subtext">
								{d3.formatPrefix(',.8', 1)(allowance) + ' allowance'}
							</div>
						) : null}
						{totalSupply ? (
							<div className="tag-subtext">
								{d3.format(',.2f')(totalSupply) + ' total supply'}
							</div>
						) : null}
					</div>
				</div>
			</div>
		</SCardAssetTag>
	);
};

const ExtendExtraDiv = (props: { accountShow: string; account: string }) => {
	const { account, accountShow } = props;
	return (
		<SCardExtendExtraDiv
			color={accountShow === 'Unknown' ? ColorStyles.TextRedAlpha : undefined}
		>
			<div className="extend-extra-wrapper">
				<div className="tag-title">Address</div>
				<a
					className="tag-content"
					href={
						'https://' +
						(__KOVAN__ ? 'kovan.' : '') +
						'etherscan.io' +
						(account ? '/address/' + account : '')
					}
					target="_blank"
				>
					{accountShow}
				</a>
			</div>
		</SCardExtendExtraDiv>
	);
};

export default class BalanceCard extends React.Component<IProps> {
	public render() {
		const { balances, account, states } = this.props;
		return (
			<SCard
				title={<SCardTitle>BALANCE</SCardTitle>}
				width="640px"
				margin="0 0 0 10px"
				extra={
					<ExtendExtraDiv accountShow={account ? account : 'Unknown'} account={account} />
				}
			>
				<SDivFlexCenter horizontal padding="0 10px">
					<BalanceInfo icon={ethIcon} name={CST.TH_ETH} value={balances.eth} />
					<BalanceInfo
						icon={duoIcon}
						name={CST.TH_DUO}
						value={balances.duo}
						allowance={balances.allowance}
					/>
					<BalanceInfo
						icon={classAIcon}
						name={CST.TH_TOKEN_A}
						value={balances.tokenA}
						totalSupply={states.totalSupplyA}
					/>
					<BalanceInfo
						icon={classBIcon}
						name={CST.TH_TOKEN_B}
						value={balances.tokenB}
						totalSupply={states.totalSupplyB}
					/>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
