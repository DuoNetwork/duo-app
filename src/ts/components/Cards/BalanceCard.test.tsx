// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { shallow } from 'enzyme';
import * as React from 'react';
import BalanceCard from './BalanceCard';

describe('BalanceCard Test', () => {
	describe('Test Snapshot', () => {
		const mobile = true;
		const account = '0x0';
		const locale = 'EN';
		const refreshBalance = jest.fn(() => 123);
		const contractAddress = {
			custodian: {
				code: 'BEETHOVEN-PPT',
				address: '0x95B3BE483e9e3685Ed631e9611b8cDba4C13641E'
			},
			aToken: {
				code: 'aETH',
				address: '0xC600fe64CDa57b607B251aa0879b8386e9FEd9f7'
			},
			bToken: {
				code: 'bETH',
				address: '0xa03b5171fE58fD2d6a018693E8D2CeD83b73ce00'
			}
		};

		it('Test Snapshot', () => {
			const wrapper = shallow(
				<BalanceCard
					contractAddress={contractAddress}
					refresh={refreshBalance}
					mobile={mobile}
					account={account}
					eth={0}
					aToken={0}
					bToken={0}
					locale={locale}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			const wrapper1 = shallow(
				<BalanceCard
					contractAddress={contractAddress}
					refresh={refreshBalance}
					mobile={false}
					account={account}
					eth={0}
					aToken={0}
					bToken={0}
					locale={locale}
				/>
			);
			expect(wrapper1).toMatchSnapshot();
		});
	});
});
