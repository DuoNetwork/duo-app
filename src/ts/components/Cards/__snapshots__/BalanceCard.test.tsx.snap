// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthCard Test User Login Test Snapshot 1`] = `
<Styled(Card)
  extra={
    <ExtendExtraDiv
      account="0x0"
      accountShow="0x0"
      locale="EN"
    />
  }
  margin="20px 0 0 0"
  title={
    <styled.div>
      BALANCE
       
      <Styled(Button)
        icon="reload"
        onClick={[MockFunction]}
      />
    </styled.div>
  }
  width="100%"
>
  <styled.div
    horizontal={false}
    padding="0 10px"
  >
    <BalanceInfo
      icon={null}
      mobile={true}
      name="ETH"
      value={0}
    />
    <BalanceInfo
      icon={null}
      mobile={true}
      name="Token A"
      value={0}
    />
    <BalanceInfo
      icon={null}
      mobile={true}
      name="Token B"
      value={0}
    />
  </styled.div>
</Styled(Card)>
`;

exports[`AuthCard Test User Login Test Snapshot 2`] = `
<Styled(Card)
  extra={
    <ExtendExtraDiv
      account="0x0"
      accountShow="0x0"
      locale="EN"
    />
  }
  margin="0 0 0 10px"
  title={
    <styled.div>
      BALANCE
       
      <Styled(Button)
        icon="reload"
        onClick={[MockFunction]}
      />
    </styled.div>
  }
  width="590px"
>
  <styled.div
    horizontal={true}
    padding="0 10px"
  >
    <BalanceInfo
      icon={null}
      mobile={false}
      name="ETH"
      value={0}
    />
    <BalanceInfo
      icon={null}
      mobile={false}
      name="Token A"
      value={0}
    />
    <BalanceInfo
      icon={null}
      mobile={false}
      name="Token B"
      value={0}
    />
  </styled.div>
</Styled(Card)>
`;
